src/main/java/com/limechain/trie/DiskTrieService.java

```diff
@@ -0,0 +1,826 @@
+
+        result.put(branchNodeFullKey, toPendingInsertUpdate(
+            null,
+            newNodePartialKey.take(branchPartialKeyLen),
+            newNodeStateVersion,
+            newNodeChildren,
+            ancestor == null
+        ));
+
+        return result;
+    }
+
+    @Nullable
+    private TrieNodeData getCachedChildAtIndex(Nibbles parentFullKey, Nibble childIndex) {
+        Optional<PendingInsertUpdate> cached = trieChanges.getChildByIndex(parentFullKey, childIndex);
+
+        if (cached.isPresent()) {
+            PendingInsertUpdate update = cached.get();
+            return new TrieNodeData(
+                update.value() == null,
+                update.partialKey(),
+                new ArrayList<>(update.childrenMerkleValues()),
+                update.value(),
+                null,
+                (byte) update.stateVersion().asInt()
+            );
+        }
+
+        return null;
+    }
+
+    public void deleteStorageNode(byte[] merkleRoot, Nibbles key) {
+        TraversalResult traversalResult = traverseTrie(merkleRoot, key);
+
+        switch (traversalResult) {
+            case TraversalResult.Found found -> {
+                TreeMap<Nibbles, PendingTrieNodeChange> executionUpdates = executeDeletion(found.getFound());
+                trieChanges.updateCache(mergeDeletionUpdatesWithTraversed(
+                    executionUpdates, traversalResult.traversedNodes));
+            }
+            case TraversalResult.NotFound ignored -> log.fine("DELETE: Node not found at key " + key);
+            case TraversalResult.Unfinished ignored ->
+                throw new IllegalStateException("Traversal result cannot be unfinished at this point in the logic");
+        }
+    }
+
+    public DeleteByPrefixResult deleteMultipleNodesByPrefix(byte[] merkleRoot, Nibbles prefix, Long limit) {
+        TraversalResult traversalResult = traverseTrie(merkleRoot, prefix);
+
+        switch (traversalResult) {
+            case TraversalResult.Found found -> {
+                AtomicInteger deleted = new AtomicInteger(0);
+
+                TraversedNode node = found.getFound();
+
+                for (Nibble nibble : Nibbles.ALL) {
+                    byte[] childMerkle = node.getChildrenMerkleValues().get(nibble.asInt());
+                    if (childMerkle != null) {
+                        Optional<PendingTrieNodeChange> recursionResult = executeRecursiveDeletion(
+                            node.getFullKey(), nibble, childMerkle, limit, deleted);
+                        if (recursionResult.isPresent()) {
+                            PendingTrieNodeChange change = recursionResult.get();
+                            node.getChildrenMerkleValues().set(nibble.asInt(),
+                                change instanceof PendingInsertUpdate c
+                                    ? c.newMerkleValue()
+                                    : null);
+                        }
+                    }
+                }
```

Comment:
When you have nested logic like that it's always good to leave 1 comment shortly explaining what's going on. Might add such a comment in several places in this file.

Reply:
I'm working on a thorough documentation of this file specifically.